### PR TITLE
Cast member pointer arguments of class_<T>::def() to member-of-T

### LINKED ISSUE #470: [BUG]: can't call exposed member of a type not known to nanobind
### Problem description

Imagine you have this class and corresponding nanobind binding for it:
```
class Example {
  public:
    void do_it();
};

nb::class_<Example>(m, "Example").def("do_it", &Example::do_it);
```
Now someone comes along and refactors Example, in a way that they think should be transparent to existing users:
```
class ExampleBase {
  public:
    void do_it();
};
class Example : public ExampleBase {};
```
The nanobind bindings above will still compile, but calling `do_it()` won't work anymore. That's because the type of `&Example::do_it` is now `void (ExampleBase::*)()`, which nanobind converts into a lambda with signature `(ExampleBase*) -> void`, which produces a TypeError when you call it, because nanobind has no idea what an `ExampleBase` is.

I think the correct solution here (which pybind11 already does) is to implicit-cast the member pointer provided to `class_<T>::def()` to be a member of `T` before passing it along to `cpp_function`. This is a safe cast that can be done at compile time as long as there's no virtual inheritance, and even properly handles the case where `T` starts at a different offset than its base (such as because `T` is polymorphic and the base is not). I'll put up a PR for it.

### Reproducible example code

```text
/* C++ */

#include <nanobind/nanobind.h>
namespace nb = nanobind;

class ExampleBase {
  public:
    void do_it();
};
class Example : public ExampleBase {};

NB_MODULE(example, m) {
    nb::class_<Example>(m, "Example")
        .def(nb::init<>())
        .def("do_it", &Example::do_it);
}

# Python

import example
ex = example.Example()
ex.do_it()
```

### PR DESCRIPTION
**Pull request by oremanj**
Fixes #470; see that issue for a more detailed explanation of the problem. This ports over pybind's solution with some C++ modernization.

### CODE REVIEW

File: include/nanobind/nb_class.h [hunk 1]
```diff
@@ -262,6 +262,37 @@ struct is_copy_constructible : std::is_copy_constructible<T> { };
...
+// we need to check `is_same` as well.
+template <typename Base, typename Derived>
+constexpr bool is_accessible_base_of_v =
+        (std::is_same_v<Base, Derived> || std::is_base_of_v<Base, Derived>)
+        && std::is_convertible_v<Derived *, Base *>;
+
+// Given a pointer to a member function, cast it to its `Derived` version.
+// Forward everything else unchanged. Analogous to method_adaptor in pybind11.
+template <typename /*Derived*/, typename F>
+auto with_self(F &&f) -> decltype(std::forward<F>(f)) {
```
[the remark this answers is not in the document]
Comment (oremanj) [reply include/nanobind/nb_class.h:276]: decltype of a function call captures the value category of the function's return type; std::forward returns either an rvalue or lvalue reference, so this function does as well. But, irrelevant with the simplification I'm making.

File: include/nanobind/nb_class.h [hunk 2]
```diff
@@ -262,6 +262,37 @@ struct is_copy_constructible : std::is_copy_constructible<T> { };
...
+
+template <typename Derived, typename Return, typename Class, typename... Args>
+auto with_self(Return (Class::*pmf)(Args...)) -> Return (Derived::*)(Args...) {
+    static_assert(is_accessible_base_of_v<Class, Derived>,
+                  "Cannot bind an inaccessible base class method; "
+                  "use a lambda definition instead");
+    return pmf;
+}
+
+template <typename Derived, typename Return, typename Class, typename... Args>
```
[the remark this answers is not in the document]
Comment (oremanj) [reply include/nanobind/nb_class.h:288]: `analyze_method` is not actually used when binding a member pointer (only a lambda); we rely on overloads of `cpp_function` and `cpp_function_def` for that, and those only exist for regular and const members. I think noexcept members are probably adding an implicit conversion to their non-noexcept equivalent, but I'm not certain of that diagnosis. In any event, irrelevant given the upcoming simplification.

**Comment by oremanj** [comment]
Thanks for encouraging me to simplify this. Please take another look; I think you'll be much happier with the second version. It should work on anything that `cpp_function` does, because it's implemented inside the relevant overloads of `cpp_function`.

**Comment by wjakob** [comment]
Yes, I am much happier with this version, thank you for making the changes. This looks good to merge to me, shall I?

**Comment by oremanj** [comment]
Go for it - thank you!

**Comment by wjakob** [comment]
Sorry, one more request: could you add a changelog entry?

**Comment by wjakob** [comment]
I think that the issue in #284 is also simplified by this commit.

**Comment by oremanj** [comment]
Added changelog entry.

**Comment by wjakob** [comment]
Thank you 👍

**Pull request merged**